File: backend/services/contestService.js
const axios = require('axios');

const COLORS = {
    leetcode: '#dc2626',
    codeforces: '#3b82f6',
    codechef: '#10b981',
    atcoder: '#6b7280',
    default: '#8b5cf6'
};

const normalizeDate = (date) => new Date(date).toISOString();

const getCodeforcesType = (name, type) => {
    const divMatch = name.match(/Div\.?\s*([0-9\+]+)/i);
    if (name.includes('Educational') && divMatch) {
        return `Edu Div ${divMatch[1]}`;
    }
    if (name.includes('Educational')) return 'Educational';
    if (divMatch) return `Div ${divMatch[1]}`;
    return type === 'CF' ? 'Division' : 'Other';
};

let cachedContests = null;
let lastFetchTime = 0;
const CACHE_DURATION = 60 * 60 * 1000; // 1 hour

const fetchContests = async () => {
    // Check Cache
    if (cachedContests && (Date.now() - lastFetchTime < CACHE_DURATION)) {
        console.log('Serving contests from cache');
        return cachedContests;
    }

    try {
        const [cfRes, atRes, lcRes, ccRes] = await Promise.allSettled([
            axios.get(process.env.CODEFORCES_API_URL, { timeout: 10000 }),
            axios.get(process.env.ATCODER_API_URL, { timeout: 10000 }),
            axios.get(process.env.LEETCODE_API_URL, { timeout: 10000 }),
            axios.get(process.env.CODECHEF_API_URL, { timeout: 10000 })
        ]);

        let contests = [];
        const now = new Date(); // Current time

        // Calculate the date range for filtering contests
        // Default range: Previous month start to end of 2 months later (approx 90 days window)
        const start = new Date(now.getFullYear(), now.getMonth() - 1, 1);
        const end = new Date(now.getFullYear(), now.getMonth() + 2, 0, 23, 59, 59);

        const isInRange = (dateStr) => {
            const date = new Date(dateStr);
            return date >= start && date <= end;
        };

        // Codeforces
        if (cfRes.status === 'fulfilled' && cfRes.value.data.status === 'OK') {
            const cfContests = cfRes.value.data.result
                .filter(c => isInRange(c.startTimeSeconds * 1000))
                .map(c => ({
                    title: c.name,
                    start: normalizeDate(c.startTimeSeconds * 1000),
                    end: normalizeDate((c.startTimeSeconds + c.durationSeconds) * 1000),
                    platform: 'Codeforces',
                    type: getCodeforcesType(c.name, c.type),
                    url: `https://codeforces.com/contests`,
                    color: COLORS.codeforces
                }));
            contests = [...contests, ...cfContests];
        }

        // AtCoder
        if (atRes.status === 'fulfilled') {
            const atContests = atRes.value.data
                .filter(c => c.start_epoch_second > 0)
                .map(c => ({ ...c, startMs: c.start_epoch_second * 1000 }))
                .filter(c => isInRange(c.startMs))
                .map(c => ({
                    title: c.title,
                    start: new Date(c.startMs).toISOString(),
                    end: new Date(c.startMs + c.duration_second * 1000).toISOString(),
                    platform: 'AtCoder',
                    type: c.id.startsWith('abc') ? 'beginner' : c.id.startsWith('arc') ? 'regular' : c.id.startsWith('agc') ? 'grand' : 'other',
                    url: `https://atcoder.jp/contests/${c.id}`,
                    color: COLORS.atcoder
                }));
            contests.push(...atContests);
        }

        // LeetCode
        if (lcRes.status === 'fulfilled' && lcRes.value.data && lcRes.value.data.allContests) {
            const lcContests = lcRes.value.data.allContests
                .filter(c => isInRange(c.startTime * 1000))
                .map(c => ({
                    title: c.title,
                    start: normalizeDate(c.startTime * 1000),
                    end: normalizeDate((c.startTime + c.duration) * 1000),
                    platform: 'LeetCode',
                    type: c.title.toLowerCase().includes('biweekly') ? 'biweekly' : 'weekly',
                    url: `https://leetcode.com/contest/${c.titleSlug}`,
                    color: COLORS.leetcode
                }));
            contests = [...contests, ...lcContests];
        }

        // CodeChef
        if (ccRes.status === 'fulfilled' && ccRes.value.data) {
            const ccContests = [
                ...(ccRes.value.data.present_contests || []),
                ...(ccRes.value.data.future_contests || [])
            ]
                .filter(c => isInRange(c.contest_start_date_iso))
                .map(c => ({
                    title: c.contest_name,
                    start: normalizeDate(c.contest_start_date_iso),
                    end: normalizeDate(c.contest_end_date_iso),
                    platform: 'CodeChef',
                    type: c.contest_code.includes('START') ? 'starters' : 'long',
                    url: `https://www.codechef.com/${c.contest_code}`,
                    color: COLORS.codechef
                }));
            contests = [...contests, ...ccContests];
        }

        contests.sort((a, b) => new Date(a.start) - new Date(b.start));

        // Update Cache
        cachedContests = contests;
        lastFetchTime = Date.now();
        console.log('Fetched fresh contests and updated cache');

        return contests;

    } catch (error) {
        console.error("Error fetching contests:", error);
        return [];
    }
};

module.exports = { fetchContests };

File: flask_based_app/services/contest_service.py
import os
import requests
from datetime import datetime, timedelta
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

COLORS = {
    'leetcode': '#f59e0b',
    'codeforces': '#3b82f6',
    'codechef': '#10b981',
    'atcoder': '#8b5cf6',
    'default': '#8b5cf6'
}

cached_contests = None
last_fetch_time = 0
CACHE_DURATION = 3600  # 1 hour in seconds

def get_codeforces_type(name, type_):
    if 'Div.' in name:
        parts = name.split('Div.')
        if len(parts) > 1:
            div = parts[1].strip().split(' ')[0]
            if 'Educational' in name:
                return f"Edu Div {div}"
            return f"Div {div}"
    if 'Educational' in name:
        return 'Educational'
    return 'Division' if type_ == 'CF' else 'Other'

def get_leetcode_type(title):
    title_lower = title.lower()
    if 'biweekly' in title_lower:
        return 'Biweekly'
    elif 'weekly' in title_lower:
        return 'Weekly'
    else:
        return 'Contest'

def get_contest_status(start_iso, end_iso):
    now = datetime.now()
    start_dt = datetime.fromisoformat(start_iso)
    end_dt = datetime.fromisoformat(end_iso)
    if now < start_dt:
        return 'upcoming'
    elif start_dt <= now < end_dt:
        return 'live'
    else:
        return 'done'

def fetch_contests():
    global cached_contests, last_fetch_time
    
    now = datetime.now()
    if cached_contests and (now.timestamp() - last_fetch_time < CACHE_DURATION):
        logger.info('Serving contests from cache')
        # Add status to cached contests
        for contest in cached_contests:
            contest['status'] = get_contest_status(contest['start'], contest['end'])
        return cached_contests

    contests = []
    
    # Calculate date range
    start_date = (now.replace(day=1) - timedelta(days=1)).replace(day=1)
    end_date = (now.replace(day=1) + timedelta(days=62)).replace(day=1) - timedelta(days=1)
    
    def is_in_range(date_obj):
        return start_date <= date_obj <= end_date

    # URLs from env
    cf_url = os.getenv('CODEFORCES_API_URL', 'https://codeforces.com/api/contest.list?gym=false')
    at_url = os.getenv('ATCODER_API_URL', 'https://kenkoooo.com/atcoder/resources/contests.json')
    lc_url = os.getenv('LEETCODE_API_URL', 'https://alfa-leetcode-api.onrender.com/contests')
    cc_url = os.getenv('CODECHEF_API_URL', 'https://www.codechef.com/api/list/contests/all')

    # ---- Fetch LeetCode ----
    lc_fetched = False

    # Attempt 1: alfa-leetcode-api proxy
    try:
        # The proxy has /contests which returns all upcoming contests
        res = requests.get(lc_url, timeout=12)
        if res.status_code == 200:
            data = res.json()
            # The response can be a dict with keys or a list depending on version
            contest_list = []
            if isinstance(data, list):
                contest_list = data
            elif isinstance(data, dict):
                # May have keys like 'topTwoContests', 'allContests', etc.
                for key in ['topTwoContests', 'allContests', 'upcoming', 'contests']:
                    if key in data and isinstance(data[key], list):
                        contest_list.extend(data[key])
                # If it's just a flat dict with contest fields, skip
                if not contest_list and data.get('title'):
                    contest_list = [data]

            for c in contest_list:
                try:
                    start_ts = c.get('startTime')
                    if start_ts:
                        start_dt = datetime.fromtimestamp(int(start_ts))
                    elif c.get('originStartTime'):
                        start_dt = datetime.fromtimestamp(int(c['originStartTime']))
                    else:
                        continue

                    duration = int(c.get('duration', 5400))  # default 1.5 hours
                    end_dt = start_dt + timedelta(seconds=duration)
                    title = c.get('title', 'LeetCode Contest')
                    slug = c.get('titleSlug') or c.get('contestSlug') or ''

                    if is_in_range(start_dt):
                        start_iso = start_dt.isoformat()
                        end_iso = end_dt.isoformat()
                        contests.append({
                            'title': title,
                            'start': start_iso,
                            'end': end_iso,
                            'platform': 'LeetCode',
                            'type': get_leetcode_type(title),
                            'url': f'https://leetcode.com/contest/{slug}' if slug else 'https://leetcode.com/contest/',
                            'color': COLORS['leetcode'],
                            'status': get_contest_status(start_iso, end_iso)
                        })
                        lc_fetched = True
                except Exception:
                    pass
    except Exception as e:
        logger.warning(f"LeetCode proxy contest fetch failed: {e}")

    # Attempt 2: Direct LeetCode GraphQL (fallback)
    if not lc_fetched:
        try:
            query = """
            query {
                allContests {
                    title
                    titleSlug
                    startTime
                    duration
                }
            }
            """
            headers = {
                'Content-Type': 'application/json',
                'Referer': 'https://leetcode.com/contest/',
                'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36'
            }
            res = requests.post(
                'https://leetcode.com/graphql',
                json={'query': query},
                headers=headers,
                timeout=10
            )
            if res.status_code == 200:
                data = res.json()
                all_contests = data.get('data', {}).get('allContests', [])
                for c in all_contests:
                    try:
                        start_dt = datetime.fromtimestamp(c['startTime'])
                        if is_in_range(start_dt):
                            end_dt = start_dt + timedelta(seconds=c['duration'])
                            title = c['title']
                            start_iso = start_dt.isoformat()
                            end_iso = end_dt.isoformat()
                            contests.append({
                                'title': title,
                                'start': start_iso,
                                'end': end_iso,
                                'platform': 'LeetCode',
                                'type': get_leetcode_type(title),
                                'url': f"https://leetcode.com/contest/{c['titleSlug']}",
                                'color': COLORS['leetcode'],
                                'status': get_contest_status(start_iso, end_iso)
                            })
                    except Exception:
                        pass
                lc_fetched = True
        except Exception as e:
            logger.warning(f"LeetCode GraphQL contest fetch failed: {e}")

    # ---- Fetch Codeforces ----
    try:
        res = requests.get(cf_url, timeout=10)
        if res.status_code == 200:
            cf_contests = res.json().get('result', [])
            for c in cf_contests:
                try:
                    if c.get('phase') == 'BEFORE':
                        start_dt = datetime.fromtimestamp(c['startTimeSeconds'])
                        if is_in_range(start_dt):
                            end_dt = start_dt + timedelta(seconds=c.get('durationSeconds', 7200))
                            start_iso = start_dt.isoformat()
                            end_iso = end_dt.isoformat()
                            contests.append({
                                'title': c['name'],
                                'start': start_iso,
                                'end': end_iso,
                                'platform': 'Codeforces',
                                'type': get_codeforces_type(c['name'], c['type']),
                                'url': f"https://codeforces.com/contests/{c['id']}",
                                'color': COLORS['codeforces'],
                                'status': get_contest_status(start_iso, end_iso)
                            })
                except Exception:
                    pass
    except Exception as e:
        logger.warning(f"Codeforces contest fetch failed: {e}")

    # ---- Fetch AtCoder ----
    try:
        res = requests.get(at_url, timeout=10)
        if res.status_code == 200:
            ac_contests = res.json()
            for c in ac_contests:
                try:
                    start_dt = datetime.fromtimestamp(c['start_epoch_second'])
                    if is_in_range(start_dt):
                        end_dt = start_dt + timedelta(seconds=c['duration_second'])
                        start_iso = start_dt.isoformat()
                        end_iso = end_dt.isoformat()
                        contests.append({
                            'title': c['title'],
                            'start': start_iso,
                            'end': end_iso,
                            'platform': 'AtCoder',
                            'type': c['rate_change'],
                            'url': f"https://atcoder.jp/contests/{c['id']}",
                            'color': COLORS['atcoder'],
                            'status': get_contest_status(start_iso, end_iso)
                        })
                except Exception:
                    pass
    except Exception as e:
        logger.warning(f"AtCoder contest fetch failed: {e}")

    # ---- Fetch CodeChef ----
    try:
        headers = {'User-Agent': 'Mozilla/5.0'}
        res = requests.get(cc_url, headers=headers, timeout=10)
        if res.status_code == 200:
            data = res.json()
            # Future contests
            for c in data.get('future_contests', []):
                try:
                    start_dt = datetime.strptime(c['contest_start_date_iso'], "%Y-%m-%dT%H:%M:%S%z").astimezone(None).replace(tzinfo=None)
                    if is_in_range(start_dt):
                        end_dt = datetime.strptime(c['contest_end_date_iso'], "%Y-%m-%dT%H:%M:%S%z").astimezone(None).replace(tzinfo=None)
                        start_iso = start_dt.isoformat()
                        end_iso = end_dt.isoformat()
                        contests.append({
                            'title': c['contest_name'],
                            'start': start_iso,
                            'end': end_iso,
                            'platform': 'CodeChef',
                            'type': 'Contest',
                            'url': f"https://www.codechef.com/{c['contest_code']}",
                            'color': COLORS['codechef'],
                            'status': get_contest_status(start_iso, end_iso)
                        })
                except Exception:
                    pass
            # Present contests
            for c in data.get('present_contests', []):
                try:
                    start_dt = datetime.strptime(c['contest_start_date_iso'], "%Y-%m-%dT%H:%M:%S%z").astimezone(None).replace(tzinfo=None)
                    if is_in_range(start_dt):
                        end_dt = datetime.strptime(c['contest_end_date_iso'], "%Y-%m-%dT%H:%M:%S%z").astimezone(None).replace(tzinfo=None)
                        start_iso = start_dt.isoformat()
                        end_iso = end_dt.isoformat()
                        contests.append({
                            'title': c['contest_name'],
                            'start': start_iso,
                            'end': end_iso,
                            'platform': 'CodeChef',
                            'type': 'Contest',
                            'url': f"https://www.codechef.com/{c['contest_code']}",
                            'color': COLORS['codechef'],
                            'status': get_contest_status(start_iso, end_iso)
                        })
                except Exception:
                    pass
    except Exception as e:
        logger.warning(f"CodeChef contest fetch failed: {e}")

    # Sort contests by start time
    contests.sort(key=lambda x: x['start'])
    
    cached_contests = contests
    last_fetch_time = now.timestamp()
    
    logger.info(f'Fetched {len(contests)} contests from APIs')
    return contests

